sql/core/src/main/scala/org/apache/spark/sql/execution/TungstenCache.scala

```diff
@@ -0,0 +1,179 @@
+                  blockSize)
+                val baos = new ByteArrayOutputStream(blockSize)
+                val compressedBaos = codec.compressedOutputStream(baos)
+                compressedBaos.write(blockArray)
+                compressedBaos.flush()
+                compressedBaos.close()
+                baos.toByteArray
+              }
+
+              // Allocate a new block with compressed byte array padded to word boundary
+              val totalRecordSize = compressedBlockArray.length + 4
+              val nearestWordBoundary =
+                ByteArrayMethods.roundNumberOfBytesToNearestWord(totalRecordSize)
+              val padding = nearestWordBoundary - totalRecordSize
+              val compressedBlock = taskMemoryManager.allocateUnchecked(totalRecordSize + padding)
+              Platform.putInt(
+                compressedBlock.getBaseObject,
+                compressedBlock.getBaseOffset,
+                padding)
+              Platform.copyMemory(
+                compressedBlockArray,
+                Platform.BYTE_ARRAY_OFFSET,
+                compressedBlock.getBaseObject,
+                compressedBlock.getBaseOffset + 4,
+                compressedBlockArray.length)
+              taskMemoryManager.freeUnchecked(block)
+              compressedBlock
+            case None => block
+          }
+        }
+
+        def hasNext: Boolean = bufferedRowIterator.hasNext
+      }
+    }.setName(compressionType + "_" + child.nodeName).persist(StorageLevel.MEMORY_ONLY)
+    // TODO(josh): is this the right name?
+
+    cachedRDD.mapPartitions { blockIterator =>
+      blockIterator.flatMap { rawBlock =>
+        // Optionally decompress block
+        val compressionCodec: Option[CompressionCodec] =
+          compressionType.map { t => CompressionCodec.createCodec(SparkEnv.get.conf, t) }
+        val block = compressionCodec match {
+          case Some(codec) =>
+            // Copy compressed block (excluding padding) to on-heap byte array
+            val padding = Platform.getInt(rawBlock.getBaseObject, rawBlock.getBaseOffset)
+            val compressedBlockArray = new Array[Byte](blockSize)
+            Platform.copyMemory(
+              rawBlock.getBaseObject,
+              rawBlock.getBaseOffset + 4,
+              compressedBlockArray,
+              Platform.BYTE_ARRAY_OFFSET,
+              rawBlock.size() - padding)
+
+            // Decompress into MemoryBlock backed by on-heap byte array
+            val decompressionStream =
+              codec.compressedInputStream(new ByteArrayInputStream(compressedBlockArray))
+            val decompressedBlock = new Array[Byte](blockSize)
+            ByteStreams.readFully(decompressionStream, decompressedBlock)
+            decompressionStream.close()
+            MemoryBlock.fromByteArray(decompressedBlock)
+          case None => rawBlock
+        }
+
+        val rows = new ArrayBuffer[UnsafeRow]()
+        var currOffset = 0
```

Comment:
I'm going to make this into a long to avoid warnings about possible overflow when comparing to the block size (which is a long).